### PR TITLE
Add some gc safety around _mysql__fetch_row

### PR DESCRIPTION
**Pull request by fried**
Users of gc.get_referrers() could cause a SystemError to be raised if this function is running in a different python thread.

We ended up doing this at Facebook just to be safe, because it was really hard figuring out that it was not a _mysql issue but usage of gc.get_referrers() in a multi-threaded environment. but I understand if you don't feel its needed.

### CODE REVIEW

**Comment by methane** [comment]
Please detailed information about why and how gc.get_referrers() is broken.

**Comment by fried** [comment]
gc.get_referrers is not broken per say, its just not safe in a multi threaded environment when c-api is used. 

We added this deflect concerns that _mysql.c was broken or python itself had a bug, when the behavior of gc module is well known and not considered a bug by the python core devs.

**Comment by methane** [comment]
I said "detailed information"!

I don't understand the problem you're saying yet.  Only information I got is it's `SystemError`.
You don't provide error message, etc.

I don't want merge such code without understanding.

**Comment by methane** [comment]
> and not considered a bug by the python core devs.

Link to bugs.python.org issue.

**Comment by fried** [comment]
```
  File "MySQLdb/cursors.py", line 203, in execute

SystemError: Objects/tupleobject.c:854: bad argument to internal function
```

Not that its very useful

**Comment by methane** [comment]
> Not that its very useful

It's very important information.  What is your exact Python version?

**Comment by methane** [comment]
OK, I understand the error is from `_PyTuple_Resize()`.

It is a dirty part of MySQLdb I dislike.
Rows should be list instead of tuple, but I couldn't change it for Django compatibility...

Anyway, thank you for reporting the issue, while I don't like this change.

File: MySQLdb/_mysql.c [hunk 1]
```diff
@@ -1284,9 +1284,15 @@ _mysql__fetch_row(
         if (!self->use)
             row = mysql_fetch_row(self->result);
         else {
+            // see: https://docs.python.org/3/library/gc.html#gc.get_referrers
+            // This function can get a reference to the tuple r, and if that
+            // code is preempted while holding a ref to r, the _PyTuple_Resize
+            // will raise a SystemError because the ref count is 2.
+            PyObject_GC_UnTrack(*r);
```
Comment (methane) [inline MySQLdb/_mysql.c]: Move these code to _mysql_ResultObject_fetch_row.

Comment (fried) [reply MySQLdb/_mysql.c]: its pretty safe everywhere else we do it in the dangerous area where we allow threads.

Comment (methane) [reply MySQLdb/_mysql.c]: It is safe untrack soon after creatin, and track again after _PyTuple_Resize().

Comment (methane) [reply MySQLdb/_mysql.c]: For example, `convert_row` may execute Python code.
Execute python code means it allows threads switching,  Calling gc module, etc.

Py_BEGIN_ALLOW_THREADS is not the only danger API.  There are many danger APIs
which can execute any Python code.  e.g. Py_DECREF, PyObject_RichCompareBool,
PyList_Append, etc.

Comment (methane) [reply MySQLdb/_mysql.c]: ping?

**Pull request merged**